src/pages/posts/comparing-rust-smart-contract-storage-approaches.md

```diff
@@ -0,0 +1,515 @@
+
+Anyhow, the interesting part is that it **needs** to be initialized. Why?
+
+Because a given set of Wasm bytes is **only deployed once**, in CosmWasm.
+
+Let me say that again; it's super cool.
+
+If you deploy the same NFT contract a second or tenth or thousandth time on CosmWasm, _no new bytes will be stored_.
+
+The on-chain storage _will not get bloated_ with a bunch of repetitive contract code.
+
+This comes at the cost of requiring contracts to be instantiated. Why? Because you might not need to deploy a contract at all. If the contract you want to "deploy" already has its bytes stored on-chain, you can find the code ID, or address, of those bytes. Then you reference that code ID when you instantiate.
+
+And anyhow, this barely counts as an added cost. Any contract of realistic size probably needs some initialization code, anyhow. The only contracts I've seen that don't require initialization are toy contracts like this incrementer.
+
+So this demo contract seems a little more complicated than the NEAR and Soroban versions, because you need to explicitly initialize the contract, setting the `counter` to `0` while you do so. NEAR and Soroban let you default `counter` to `0`, rather than explicitly setting it. But they also both require duplicating contract bytes on-chain over and over. Score one for CosmWasm, imho.
+
+### Errors
+
+One last thing to get out of the way we focus in on storage manipulation.
+
+The CosmWasm template comes with a bunch of error handling stuff. Most notably:
+
+```rust
+#[derive(Error, Debug)]
+pub enum ContractError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+}
+```
+
+You can add your own errors to this enum, and include user-facing messages using [thiserror](https://docs.rs/thiserror/1.0.21/thiserror/).
+
+Then it gets returned throughout calls to the contract in all those `Result`s. For example:
+
+```rust
+pub fn execute(…) -> Result<Response, ContractError> {…}
+```
+
+This is another area where it kinda seems like CosmWasm requires a lot of ceremony, but it's actually just establishing firm conventions (and great usability) for something that all realistic contracts will need. Once you get your bearings, Soroban encourages a [similar approach to CosmWasm](https://soroban.stellar.org/docs/examples/errors), but less baked-in. And NEAR doesn't have any conventions around this; most people [use the `require!` macro](https://docs.near.org/sdk/rust/best-practices) and throw user-facing messages right in the contract logic. Maybe that's fine?
```

Comment:
> Maybe it's fine?

It means that consumers of the contract call won't know what possible errors could happen or which ones did. It's been fine, but having more structured errors with a known set is a big win for automating generating exceptions in JS, etc.

Reply:
How's this sound, @willemneal?

```suggestion
This is another area where it kinda seems like CosmWasm requires a lot of ceremony, but it's actually just establishing firm conventions (and great usability) for something that all realistic contracts will need. Once you get your bearings, Soroban encourages a [similar approach to CosmWasm](https://soroban.stellar.org/docs/examples/errors), but less baked-in. And NEAR doesn't have any conventions around this; most people [use the `require!` macro](https://docs.near.org/sdk/rust/best-practices) and throw user-facing messages right in the contract logic. This has worked alright to a point, but makes it hard for consumers of NEAR contracts and builders of NEAR tooling to know what possible errors a contract can throw.
```

Reply:
Yep!